zcash_primitives/src/merkle_tree.rs

```diff
@@ -44,590 +23,307 @@ pub trait HashSer {
+) -> io::Result<()> {
+    Optional::write(writer, frontier.value(), write_nonempty_frontier_v1)
+}
 
-        PathFiller {
-            queue: self.filled.iter().cloned().chain(cursor_root).collect(),
-        }
+#[allow(clippy::redundant_closure)]
+pub fn read_frontier_v1<H: HashSer + Clone, R: Read>(reader: R) -> io::Result<Frontier<H, 32>> {
+    match Optional::read(reader, read_nonempty_frontier_v1)? {
+        None => Ok(Frontier::empty()),
+        Some(f) => Frontier::try_from(f).map_err(|err| {
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("Parsing resulted in an invalid Merkle frontier: {:?}", err),
+            )
+        }),
     }
+}
 
-    /// Finds the next "depth" of an unfilled subtree.
-    fn next_depth(&self) -> usize {
-        let mut skip = self.filled.len();
-
-        if self.tree.left.is_none() {
-            if skip > 0 {
-                skip -= 1;
-            } else {
-                return 0;
-            }
-        }
-
-        if self.tree.right.is_none() {
-            if skip > 0 {
-                skip -= 1;
-            } else {
-                return 0;
-            }
-        }
+/// Reads a legacy `CommitmentTree` from its serialized form.
+pub fn read_commitment_tree<Node: HashSer, R: Read, const DEPTH: u8>(
+    mut reader: R,
+) -> io::Result<CommitmentTree<Node, DEPTH>> {
+    let left = Optional::read(&mut reader, Node::read)?;
+    let right = Optional::read(&mut reader, Node::read)?;
+    let parents = Vector::read(&mut reader, |r| Optional::read(r, Node::read))?;
+
+    CommitmentTree::from_parts(left, right, parents).map_err(|_| {
+        io::Error::new(
+            io::ErrorKind::InvalidData,
+            "parents vector exceeded tree depth",
+        )
+    })
+}
 
-        let mut d = 1;
-        for p in &self.tree.parents {
-            if p.is_none() {
-                if skip > 0 {
-                    skip -= 1;
-                } else {
-                    return d;
-                }
-            }
-            d += 1;
-        }
+/// Serializes a legacy `CommitmentTree` as an array of bytes.
+pub fn write_commitment_tree<Node: HashSer, W: Write, const DEPTH: u8>(
+    tree: &CommitmentTree<Node, DEPTH>,
+    mut writer: W,
+) -> io::Result<()> {
+    Optional::write(&mut writer, tree.left().as_ref(), |w, n| n.write(w))?;
+    Optional::write(&mut writer, tree.right().as_ref(), |w, n| n.write(w))?;
+    Vector::write(&mut writer, tree.parents(), |w, e| {
+        Optional::write(w, e.as_ref(), |w, n| n.write(w))
+    })
+}
 
-        d + skip
-    }
+/// Reads an `IncrementalWitness` from its serialized form.
+#[allow(clippy::redundant_closure)]
+pub fn read_incremental_witness<Node: HashSer, R: Read, const DEPTH: u8>(
+    mut reader: R,
+) -> io::Result<IncrementalWitness<Node, DEPTH>> {
+    let tree = read_commitment_tree(&mut reader)?;
+    let filled = Vector::read(&mut reader, |r| Node::read(r))?;
```

Comment:
```suggestion
    let filled = Vector::read(&mut reader, Node::read)?;
```
Why not this rather than `#[allow(clippy::redundant_closure)]`?

Reply:
That doesn't actually compile, because of some confusion related to the polymorphic type.

Reply:
```
error: implementation of `FnOnce` is not general enough
   --> zcash_primitives/src/merkle_tree.rs:127:22
    |
127 |     let mut ommers = Vector::read(&mut reader, H::read)?;
    |                      ^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
    |
    = note: `fn(&'2 mut &mut R) -> Result<H, std::io::Error> {<H as HashSer>::read::<&'2 mut &mut R>}` must implement `FnOnce<(&'1 mut &mut R,)>`, for any lifetime `'1`...
    = note: ...but it actually implements `FnOnce<(&'2 mut &mut R,)>`, for some specific lifetime `'2`

error: implementation of `FnOnce` is not general enough
   --> zcash_primitives/src/merkle_tree.rs:199:18
    |
199 |     let filled = Vector::read(&mut reader, Node::read)?;
    |                  ^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
    |
    = note: `fn(&'2 mut &mut R) -> Result<Node, std::io::Error> {<Node as HashSer>::read::<&'2 mut &mut R>}` must implement `FnOnce<(&'1 mut &mut R,)>`, for any lifetime `'1`...
    = note: ...but it actually implements `FnOnce<(&'2 mut &mut R,)>`, for some specific lifetime `'2`

error: could not compile `zcash_primitives` due to 2 previous errors
```